### PR TITLE
recently viewed products

### PR DESCRIPTION
**Pull request by DilbaraAsanalieva**

### CODE REVIEW

**Review by beksultancs — CHANGES_REQUESTED** [review]
Don't like logic from find recently viewed products

File: src/main/java/kg/gadgetarium/gadgetarium/db/model/Product.java [hunk 1]
```diff
@@ -45,13 +45,18 @@ public class Product {
...
+    private List<User> users;
+    @ManyToMany
     private Set<User> userCart;
     @ManyToMany()
     private List<User> wishListUsers;
-
     @ManyToMany(cascade = CascadeType.DETACH)
     private Set<User> uc;
-
+    @ManyToMany
```
Comment (UlanKubanychbekov) [inline src/main/java/kg/gadgetarium/gadgetarium/db/model/Product.java:55]: OneToMany

Comment (beksultancs) [reply src/main/java/kg/gadgetarium/gadgetarium/db/model/Product.java:55]: berki jakty ManyToMany kylbaiby?

Comment (UlanKubanychbekov) [reply src/main/java/kg/gadgetarium/gadgetarium/db/model/Product.java:55]: бир юзердин несколько просмотренные товары болотта

File: src/main/java/kg/gadgetarium/gadgetarium/apis/RecentlyViewedProductsApi.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+import java.util.List;
+
+@RestController
+@RequestMapping("api/user/viewed/")
+@RequiredArgsConstructor
+@CrossOrigin
+@Tag(name = "Recently Viewed API")
+@PreAuthorize("hasAnyAuthority('USER')")
+public class RecentlyViewedProductsApi {
```
Comment (peaksoft-school) [inline src/main/java/kg/gadgetarium/gadgetarium/apis/RecentlyViewedProductsApi.java:20]: Why this is a separate class for one endpoint?

File: src/main/java/kg/gadgetarium/gadgetarium/apis/RecentlyViewedProductsApi.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    private final RecentlyViewedProductsService recentlyViewedProductsService;
+
+    @Operation(summary = "All products", description = "User can see all recently viewed products")
+    @GetMapping("{userId}")
+    public List<ProductRecentlyViewedResponse> getAllViews(@PathVariable Long userId){
```
Comment (peaksoft-school) [inline src/main/java/kg/gadgetarium/gadgetarium/apis/RecentlyViewedProductsApi.java:26]: We can add it to the Product API

File: src/main/java/kg/gadgetarium/gadgetarium/db/service/RecentlyViewedProductsService.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                        productRecentlyViewedResponse.setAfterDiscount(product.getPrice() * product.getDiscount().getPercent() / 100);
+                    }
+                }
+                return productRecentlyViewedResponses;
+            }
+        }
+            return null;
+    }
+}
+
```
Comment (peaksoft-school) [inline src/main/java/kg/gadgetarium/gadgetarium/db/service/RecentlyViewedProductsService.java:61]: Unnecessary lines from 61 to 76

**Review by peaksoft-school — COMMENTED** [review]
Commented 14.09.22 at 12 pm

File: src/main/java/kg/gadgetarium/gadgetarium/apis/ProductApi.java [hunk 5]
```diff
@@ -125,4 +128,17 @@ public ProductResponse getInnerPage(@PathVariable Long productId, @RequestParam
...
         return productService.getAllProductDetails();
     }
+
+    @PreAuthorize("hasAnyAuthority('USER')")
+    @Operation(summary = "All products", description = "User can see all recently viewed products")
+    @GetMapping("/viewed/products")
+    public List<ProductRecentlyViewedResponse> getAllViews(){
+        return recentlyViewedProductsService.findAllProductsByUserId();
+    }
+    @PreAuthorize("hasAnyAuthority('USER')")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/gadgetarium/gadgetarium/apis/ProductApi.java:138]: Pls write in the next line

File: src/main/java/kg/gadgetarium/gadgetarium/apis/ProductApi.java [hunk 6]
```diff
@@ -125,4 +128,17 @@ public ProductResponse getInnerPage(@PathVariable Long productId, @RequestParam
...
+    @Operation(summary = "read unread messages", description = "admin click on the button")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/gadgetarium/gadgetarium/apis/ProductApi.java:139]: read, admin "r, a" should be capital

File: src/main/java/kg/gadgetarium/gadgetarium/apis/ProductApi.java [hunk 7]
```diff
@@ -125,4 +128,17 @@ public ProductResponse getInnerPage(@PathVariable Long productId, @RequestParam
...
+    @PostMapping("viewed/{productId}")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/gadgetarium/gadgetarium/apis/ProductApi.java:140]: Should be a post method?

File: src/main/java/kg/gadgetarium/gadgetarium/db/model/ProductResponseStatus.java [hunk 8]
```diff
@@ -0,0 +1,6 @@
+package kg.gadgetarium.gadgetarium.db.model;
+
+public enum ProductResponseStatus {
+    VIEWED,
```
Comment (ChyngyzIshen) [inline src/main/java/kg/gadgetarium/gadgetarium/db/model/ProductResponseStatus.java:4]: Pls write in the next line

File: src/main/java/kg/gadgetarium/gadgetarium/db/model/User.java [hunk 9]
```diff
@@ -73,6 +73,15 @@ public void delete(Product product){
     public void deleteAll(List<Product> products) {
         this.wishLists.removeAll(products);
     }
+    public void addRecentlyProduct(Product product){
+        this.lastViewed.add(product);
+    }
+    public void deleteRecentlyViewed(Product product){
```
Comment (ChyngyzIshen) [inline src/main/java/kg/gadgetarium/gadgetarium/db/model/User.java:79]: Pls write in the next line

File: src/main/java/kg/gadgetarium/gadgetarium/db/model/User.java [hunk 10]
```diff
@@ -73,6 +73,15 @@ public void delete(Product product){
     public void deleteAll(List<Product> products) {
         this.wishLists.removeAll(products);
     }
+    public void addRecentlyProduct(Product product){
```
Comment (ChyngyzIshen) [inline src/main/java/kg/gadgetarium/gadgetarium/db/model/User.java:76]: Pls write in the next line

File: src/main/java/kg/gadgetarium/gadgetarium/db/model/User.java [hunk 11]
```diff
@@ -73,6 +73,15 @@ public void delete(Product product){
...
+        this.lastViewed.add(product);
+    }
+    public void deleteRecentlyViewed(Product product){
+        this.lastViewed.remove(product);
+    }
+    public void deleteAllRecentlyViewed(List<Product> products){
```
Comment (ChyngyzIshen) [inline src/main/java/kg/gadgetarium/gadgetarium/db/model/User.java:82]: Pls write in the next line

File: src/main/java/kg/gadgetarium/gadgetarium/db/repositories/ProductRepository.java [hunk 12]
```diff
@@ -44,6 +44,7 @@ Set<Product> findAllWithFilter(String search,
 
     @Query("select p from Product p where(select u from User u where u.id=:userId) member of p.userComparison ")
     List<Product> findAllProductsById(Long userId);
-
+    @Query("select p from Product p where(select u from User u where u.id=:userId) member of p.recentlyViewedUsers")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/gadgetarium/gadgetarium/db/repositories/ProductRepository.java:47]: Pls write in the next line

**Review by ChyngyzIshen — COMMENTED** [review]
Commented 19.09.2022 at 1.40 pm

**Pull request merged**